### PR TITLE
CoAP fixes and improvements

### PR DESCRIPTION
**Pull request by nfi**
This PR is a set of CoAP bug fixes and small feature improvements. They were developed during our work with a Contiki implementation of LWM2M and IPSO Objects that was interop tested during the IPSO Interop in May 2015.

Fixes and improvements include:
* support for observe on sub-resources
* fixed URI to always be NULL terminated for observe
* more defensive programming for bad arguments and buffer overruns

This work has been performed as a part of an open source IoT edge project and is contributed by Yanzi Networks AB and SICS Swedish ICT. Intel Corporation is a sponsor of the open source IoT edge.

### CODE REVIEW

**Comment by nfi** [comment]
I have updated the PR based on the comments and rebased on top of upstream master.

File: examples/er-rest-example/project-conf.h [hunk 1]
```diff
@@ -93,6 +96,10 @@
 #undef COAP_PROXY_OPTION_PROCESSING
 #define COAP_PROXY_OPTION_PROCESSING   0
 
+/* Turn of DAO ACK to make code smaller */
+#undef RPL_CONF_WITH_DAO_ACK
+#define RPL_CONF_WITH_DAO_ACK          0
```
Comment (simonduq) [inline examples/er-rest-example/project-conf.h]: this flag does not exist in upstream Contiki

File: examples/er-rest-example/project-conf.h [hunk 2]
```diff
@@ -60,6 +60,9 @@
 #undef NETSTACK_CONF_RDC
 #define NETSTACK_CONF_RDC              nullrdc_driver
 
+#undef RPL_CONF_MAX_DAG_PER_INSTANCE
+#define RPL_CONF_MAX_DAG_PER_INSTANCE     1
```
Comment (simonduq) [inline examples/er-rest-example/project-conf.h]: why?

File: apps/er-coap/er-coap.h [hunk 3]
```diff
@@ -156,18 +156,18 @@ typedef struct {
...
 #define COAP_SERIALIZE_BLOCK_OPTION(number, field, text) \
   if(IS_OPTION(coap_pkt, number)) \
   { \
-    PRINTF(text " [%lu%s (%u B/blk)]\n", coap_pkt->field##_num, coap_pkt->field##_more ? "+" : "", coap_pkt->field##_size); \
+    PRINTF(text " [%lu%s (%u B/blk)]\n", (unsigned long)coap_pkt->field##_num, coap_pkt->field##_more ? "+" : "", coap_pkt->field##_size); \
     uint32_t block = coap_pkt->field##_num << 4; \
     if(coap_pkt->field##_more) { block |= 0x8; } \
     block |= 0xF & coap_log_2(coap_pkt->field##_size / 16); \
-    PRINTF(text " encoded: 0x%lX\n", block); \
+    PRINTF(text " encoded: 0x%lX\n", (unsigned long)block);		\
```
Comment (simonduq) [inline apps/er-coap/er-coap.h]: isn't `%lX` signed?

**Comment by simonduq** [comment]
Very nice cleanup, and this also happens to fix https://github.com/contiki-os/contiki/pull/1250.
Haven't tested the new observe for sub-resources but the feature was definitely missing (and I trust @nfi this works).
:+1:

File: examples/er-rest-example/project-conf.h (hunk 1 above)
Comment (nfi) [inline examples/er-rest-example/project-conf.h]: This is a mistake. It is a change from our work with DAO ACK for scalability and should not be included in this PR. I will remove it. Thanks!

File: examples/er-rest-example/project-conf.h (hunk 2 above)
Comment (nfi) [inline examples/er-rest-example/project-conf.h]: This was added to reduce the code size so it could be compiled on Tmote Sky. Since I think the Travis test now compiles for WiSMote it might no longer be needed. I will remove it.

File: apps/er-coap/er-coap.h (hunk 3 above)
Comment (nfi) [inline apps/er-coap/er-coap.h]: I think %x always prints the value as unsigned. The only difference between %x and %X is that one prints lowercase and the other uppercase.

**Comment by nfi** [comment]
Thanks for your comments! I will remove the unnecessary code to reduce the example code size and do another pass through the casts in the debug output. I will let you know when it has been done.

File: apps/er-coap/er-coap.h (hunk 3 above)
Comment (simonduq) [inline apps/er-coap/er-coap.h]: right, my mistake!

**Comment by nfi** [comment]
I have updated the PR based on the comments.

File: apps/er-coap/er-coap-observe.c [hunk 4]
```diff
@@ -58,9 +58,9 @@ LIST(observers_list);
 /*---------------------------------------------------------------------------*/
 /*- Internal API ------------------------------------------------------------*/
 /*---------------------------------------------------------------------------*/
-coap_observer_t *
+static coap_observer_t *
```
Comment (nvt) [inline apps/er-coap/er-coap-observe.c]: When changing the signature of this function, the module prefix "coap_" should be removed. Furthermore, is there a need to not have the <code>const</code> qualifier set for the addr variable (and possibly doing the same for functions called from <code>coap_add_observer</code> that use <code>addr</code>?

Comment (nfi) [inline apps/er-coap/er-coap-observe.c]: I have renamed the function as suggested. It would not be a problem to change the ```addr``` argument to be ```const``` but then the other functions in the CoAP API should also be updated in the same manner. I completely agree that this would be nice but is this something we should do in this PR or leave it for another PR?

Comment (simonduq) [inline apps/er-coap/er-coap-observe.c]: I agree using `const` would be best but this seems pretty orthogonal to this PR, so I'd have it in another one. This PR is a nice cleanup and solved a known bug (https://github.com/contiki-os/contiki/pull/1250) so I'd focus on getting it in first.

Comment (nvt) [inline apps/er-coap/er-coap-observe.c]: Yes, the work on <code>const</code> is not needed for this PR to be merged.

**Comment by nvt** [comment]
:+1:
